From b162fd33b82ed7b52aa59c113be8878fc2074997 Mon Sep 17 00:00:00 2001
From: npk <npk@fedora-desktop.lan>
Date: Mon, 29 Dec 2025 04:10:05 +0100
Subject: [PATCH] drm/edid: parse HDMI VRR information

Adds a struct to hold HDMI VRR information. supported is an
additional property enabling simpler parsing in consumers and incorporates
logic commonly used by drivers to assess VRR viability.

Signed-off-by: Tomasz Pakula <tomasz.pakula.mail@gmail.com>
Tested-by: Bernhard Berger <bernhard.berger@gmail.com>
---
 drivers/gpu/drm/drm_edid.c  | 20 ++++++++++++++++++--
 include/drm/drm_connector.h | 20 ++++++++++++++++++++
 2 files changed, 38 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/drm_edid.c b/drivers/gpu/drm/drm_edid.c
index 8d56dd7..82655d9 100644
--- a/drivers/gpu/drm/drm_edid.c
+++ b/drivers/gpu/drm/drm_edid.c
@@ -6149,6 +6149,19 @@ static void drm_parse_ycbcr420_deep_color_info(struct drm_connector *connector,
 	hdmi->y420_dc_modes = dc_mask;
 }
 
+static void drm_parse_vrr_info(struct drm_connector *connector, const u8 *db)
+{
+	struct drm_hdmi_vrr_cap *vrr = &connector->display_info.hdmi.vrr_cap;
+
+	vrr->supported = false;
+	if (!(cea_db_payload_len(db) >= 10) || !db[9] || !db[10])
+		return;
+
+	vrr->min_hz = db[9] & DRM_EDID_VRR_MIN_MASK;
+	vrr->max_hz = (db[9] & DRM_EDID_VRR_MAX_UPPER_MASK) << 2 | db[10];
+	vrr->supported = (vrr->min_hz > 0) && (vrr->max_hz - vrr->min_hz > 10);
+}
+
 static void drm_parse_dsc_info(struct drm_hdmi_dsc_cap *hdmi_dsc,
 			       const u8 *hf_scds)
 {
@@ -6278,6 +6291,8 @@ static void drm_parse_hdmi_forum_scds(struct drm_connector *connector,
 	if (cea_db_payload_len(hf_scds) >= 8 && hf_scds[8])
 		hdmi->allm_supported = hf_scds[8] & DRM_EDID_ALLM;
 
+	drm_parse_vrr_info(connector, hf_scds);
+
 	if (cea_db_payload_len(hf_scds) >= 11 && hf_scds[11]) {
 		drm_parse_dsc_info(hdmi_dsc, hf_scds);
 		dsc_support = true;
@@ -6288,8 +6303,9 @@ static void drm_parse_hdmi_forum_scds(struct drm_connector *connector,
 		    connector->base.id, connector->name,
 		    max_tmds_clock, str_yes_no(max_frl_rate), str_yes_no(dsc_support));
 	drm_dbg_kms(connector->dev,
-		    "[CONNECTOR:%d:%s] ALLM support: %s\n",
-		    connector->base.id, connector->name, str_yes_no(hdmi->allm_supported));
+		    "[CONNECTOR:%d:%s] ALLM support: %s, VRR support: %s, VRR min Hz: %u, VRR max Hz: %u\n",
+		    connector->base.id, connector->name, str_yes_no(hdmi->allm_supported),
+		    str_yes_no(hdmi->vrr_cap.supported), hdmi->vrr_cap.min_hz, hdmi->vrr_cap.max_hz);
 }
 
 static void drm_parse_hdmi_deep_color_info(struct drm_connector *connector,
diff --git a/include/drm/drm_connector.h b/include/drm/drm_connector.h
index ac84884..5f95df2 100644
--- a/include/drm/drm_connector.h
+++ b/include/drm/drm_connector.h
@@ -254,6 +254,23 @@ struct drm_scdc {
 	struct drm_scrambling scrambling;
 };
 
+/**
+ * struct drm_hdmi_vrr_cap - Information about VRR capabilities of a HDMI sink
+ *
+ * Describes the VRR support provided by HDMI 2.1 sink. The information is
+ * fetched from additional HFVSDB blocks defined for HDMI 2.1.
+ */
+struct drm_hdmi_vrr_cap {
+	/** @supported: flag for vrr support by sink */
+	bool supported;
+
+	/** @min_hz: minimum supported variable refresh rate */
+	u16 min_hz;
+
+	/** @max_hz: maximum supported variable refresh rate */
+	u16 max_hz;
+};
+
 /**
  * struct drm_hdmi_dsc_cap - DSC capabilities of HDMI sink
  *
@@ -333,6 +350,9 @@ struct drm_hdmi_info {
 	/** @allm_supported: flag for auto low latency mode support by sink */
 	bool allm_supported;
 
+	/** @vrr_cap: VRR capabilities of the sink */
+	struct drm_hdmi_vrr_cap vrr_cap;
+
 	/** @dsc_cap: DSC capabilities of the sink */
 	struct drm_hdmi_dsc_cap dsc_cap;
 };
-- 
2.52.0

